derivable_impls false positive on array sizes that do not implement Default

<!--
Thank you for filing a bug report! 🐛 Please provide a short summary of the bug,
along with any information you feel relevant to replicating the bug.
-->
Lint name: derivable_impls

I have the following code:

```rust
struct RCDeserialize {
  pass2_buffer: [u8; 68],
}

impl Default for RCDeserialize {
  fn default() -> Self {
    RCDeserialize {
      pass2_buffer: [0; 68],
    }
  }
}
```

I receive this warning:

```
warning: this `impl` can be derived
   --> src/rate.rs:466:1
    |
466 | / impl Default for RCDeserialize {
467 | |   fn default() -> Self {
468 | |     RCDeserialize {
469 | |       pass2_buffer: [0; 68],
...   |
473 | |   }
474 | | }
    | |_^
    |
    = help: try annotating `rate::RCDeserialize` with `#[derive(Default)]`
    = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#derivable_impls
```

I try the suggestion:

```rust
#[derive(Default)]
struct RCDeserialize {
  pass2_buffer: [u8; 68],
}
```

I receive the following compiler error:

```
error[E0277]: the trait bound `[u8; 68]: std::default::Default` is not satisfied
   --> src/rate.rs:464:3
    |
456 | #[derive(Default)]
    |          ------- in this derive macro expansion
...
464 |   pass2_buffer: [u8; 68],
    |   ^^^^^^^^^^^^^^^^^^^^^^ the trait `std::default::Default` is not implemented for `[u8; 68]`
    |
    = help: the following implementations were found:
              <&[T] as std::default::Default>
              <&mut [T] as std::default::Default>
              <[T; 0] as std::default::Default>
              <[T; 10] as std::default::Default>
            and 31 others
note: required by `std::default::Default::default`
```

### Meta

**Rust version (`rustc -Vv`):**
```
rustc 1.57.0-nightly (9dd4ce80f 2021-09-17)
binary: rustc
commit-hash: 9dd4ce80fb01d1ff5cb5002f08b7b3847b59e664
commit-date: 2021-09-17
host: x86_64-unknown-linux-gnu
release: 1.57.0-nightly
LLVM version: 13.0.0
```

@rustbot label +I-suggestion-causes-error

--- a/clippy_utils/src/lib.rs
+++ b/clippy_utils/src/lib.rs
@@ -684,7 +684,17 @@ pub fn is_default_equivalent(cx: &LateContext<'_>, e: &Expr<'_>) -> bool {
             _ => false,
         },
         ExprKind::Tup(items) | ExprKind::Array(items) => items.iter().all(|x| is_default_equivalent(cx, x)),
-        ExprKind::Repeat(x, _) => is_default_equivalent(cx, x),
+        ExprKind::Repeat(x, y) => if_chain! {
+            if let ExprKind::Lit(ref const_lit) = cx.tcx.hir().body(y.body).value.kind;
+            if let LitKind::Int(v, _) = const_lit.node;
+            if v <= 32 && is_default_equivalent(cx, x);
+            then {
+                true
+            }
+            else {
+                false
+            }
+        },
         ExprKind::Call(repl_func, _) => if_chain! {
             if let ExprKind::Path(ref repl_func_qpath) = repl_func.kind;
             if let Some(repl_def_id) = cx.qpath_res(repl_func_qpath, repl_func.hir_id).opt_def_id();

--- a/tests/ui/derivable_impls.rs
+++ b/tests/ui/derivable_impls.rs
@@ -207,4 +207,24 @@ impl Default for Color2 {
     }
 }
 
+pub struct RepeatDefault1 {
+    a: [i8; 32],
+}
+
+impl Default for RepeatDefault1 {
+    fn default() -> Self {
+        RepeatDefault1 { a: [0; 32] }
+    }
+}
+
+pub struct RepeatDefault2 {
+    a: [i8; 33],
+}
+
+impl Default for RepeatDefault2 {
+    fn default() -> Self {
+        RepeatDefault2 { a: [0; 33] }
+    }
+}
+
 fn main() {}

--- a/tests/ui/derivable_impls.stderr
+++ b/tests/ui/derivable_impls.stderr
@@ -73,5 +73,17 @@ LL | | }
    |
    = help: try annotating `WithoutSelfParan` with `#[derive(Default)]`
 
-error: aborting due to 6 previous errors
+error: this `impl` can be derived
+  --> $DIR/derivable_impls.rs:214:1
+   |
+LL | / impl Default for RepeatDefault1 {
+LL | |     fn default() -> Self {
+LL | |         RepeatDefault1 { a: [0; 32] }
+LL | |     }
+LL | | }
+   | |_^
+   |
+   = help: try annotating `RepeatDefault1` with `#[derive(Default)]`
+
+error: aborting due to 7 previous errors
 

